### PR TITLE
Feature/shooter.v2

### PR DESCRIPTION
**Pull request by KenyonSchep**
Basic subsystem with shoot command, falcons, pistons, and driver station buttons
PTAL

### CODE REVIEW

File: src/main/include/subsystems/Shooter.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <frc2/command/SubsystemBase.h>
+#include <ctre/phoenix/motorcontrol/can/WPI_TalonFX.h>
+#include "frc/WPILib.h"
```
Comment (duane701) [inline src/main/include/subsystems/Shooter.h]: Remove WPILib.h. Add specific includes. There might not be a lot to add.

File: src/main/include/subsystems/Shooter.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "ctre/Phoenix.h"
```
Comment (duane701) [inline src/main/include/subsystems/Shooter.h]: Remove Phoenix.h. Add specific includes. There might not be a lot to add.

File: src/main/cpp/commands/Shoot.cpp [hunk 3]
```diff
@@ -7,13 +7,9 @@
...
 
-Shoot::Shoot( 
-   std::function<double()> shooter1
-,  std::function<double()> shooter2
-,  Shooter& mShooter
-,  double speed)
-: mshooter1(shooter1)
-, mshooter2(shooter2)
+Shoot::Shoot(Shooter& mShooter,  double speed)
+: mshooter1(mshooter1)
```
Comment (duane701) [inline src/main/cpp/commands/Shoot.cpp]: Seems like this is fixed later on. `cause this doesn't work.

File: src/main/cpp/subsystems/Shooter.cpp [hunk 4]
```diff
@@ -9,10 +9,14 @@
...
     const wpi::Twine& name,
-    WPI_TalonFX& shooter1,
-    WPI_TalonFX& shooter2)
-: mshooter1{shooter1}
-, mshooter2{shooter2}
+    WPI_TalonFX& shooterleft,
+    WPI_TalonFX& shooterright,
+    frc::DoubleSolenoid& hood,
+    frc::DoubleSolenoid& latch)
+: mshooterleft{shooterleft}
```
Comment (duane701) [inline src/main/cpp/subsystems/Shooter.cpp:31]: This is better.

File: src/main/cpp/subsystems/Shooter.cpp [hunk 5]
```diff
@@ -23,7 +27,7 @@ double Shooter::MotorTopRPM()
   constexpr double TPR {2048};
```
Comment (duane701) [inline src/main/cpp/subsystems/Shooter.cpp]: Constants are start with `k`. What is TPR? Let's limit abbreviations. Please write out what LTR is.

File: src/main/include/commands/HoodOutFull.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+class HoodOutFull
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 HoodOutFull> {
+ public:
+  HoodOutFull();
+
+  void Initialize() override;
+
+  private:
+  Shooter* mShooter;
```
Comment (duane701) [inline src/main/include/commands/HoodOutFull.h]: Use a reference for subsystems

```suggestion
  Shooter& mShooter;
```

File: src/main/include/commands/HoodOutFull.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <frc2/command/CommandHelper.h>
+#include <frc2/command/InstantCommand.h>
+#include <subsystems/Shooter.h>
+
+class HoodOutFull
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 HoodOutFull> {
+ public:
+  HoodOutFull();
```
Comment (duane701) [inline src/main/include/commands/HoodOutFull.h]: Need to construct commands with subsystems

```suggestion
  HoodOutFull(Shooter& shooter);
```

File: src/main/cpp/commands/HoodOutFull.cpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "commands/HoodOutFull.h"
+
+// NOTE:  Consider using this command inline, rather than writing a subclass.
+// For more information, see:
+// https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html
+HoodOutFull::HoodOutFull() {
```
Comment (duane701) [inline src/main/cpp/commands/HoodOutFull.cpp]: Construct with a subsystem

```suggestion
HoodOutFull::HoodOutFull(Shooter& shooter)
: mShooter(shooter)
 {
```

File: src/main/cpp/commands/HoodRetract.cpp [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "commands/HoodRetract.h"
+
+// NOTE:  Consider using this command inline, rather than writing a subclass.
+// For more information, see:
+// https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html
+HoodRetract::HoodRetract() {
```
Comment (duane701) [inline src/main/cpp/commands/HoodRetract.cpp]: Same comments as HoodOutFull.

File: src/main/include/commands/HoodRetract.h [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <frc2/command/CommandHelper.h>
+#include <frc2/command/InstantCommand.h>
+#include <subsystems/Shooter.h>
+
+class HoodRetract
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 HoodRetract> {
+ public:
+  HoodRetract();
```
Comment (duane701) [inline src/main/include/commands/HoodRetract.h]: Same comments as `HoodOutFull`.

File: src/main/cpp/commands/LatchDisengage.cpp [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "commands/LatchDisengage.h"
+
+// NOTE:  Consider using this command inline, rather than writing a subclass.
+// For more information, see:
+// https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html
+LatchDisengage::LatchDisengage() {
```
Comment (duane701) [inline src/main/cpp/commands/LatchDisengage.cpp]: See previous comments about subsystems and commands. Commands need subsystems by reference

```suggestion
LatchDisengage::LatchDisengage(Shooter& shooter) {
```

File: src/main/cpp/subsystems/Shooter.cpp [hunk 12]
```diff
@@ -9,6 +9,8 @@
 
 constexpr frc::DoubleSolenoid::Value kHoodOutFull {frc::DoubleSolenoid::kForward};
 constexpr frc::DoubleSolenoid::Value kHoodRetract {frc::DoubleSolenoid::kReverse};
+constexpr frc::DoubleSolenoid::Value kLatchEngage {frc::DoubleSolenoid::kForward};
+constexpr frc::DoubleSolenoid::Value kLatchDisengage {frc::DoubleSolenoid::kReverse};
```
Comment (duane701) [inline src/main/cpp/subsystems/Shooter.cpp:20]: Put the `constexpr`'s all together into a `namespace {` block.

File: src/main/include/commands/LatchDisengage.h [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <frc2/command/CommandHelper.h>
+#include <frc2/command/InstantCommand.h>
+#include <subsystems/Shooter.h>
+
+class LatchDisengage
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 LatchDisengage> {
+ public:
+  LatchDisengage();
```
Comment (duane701) [inline src/main/include/commands/LatchDisengage.h]: More subsystems

File: src/main/include/commands/LatchEngage.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+class LatchEngage
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 LatchEngage> {
+ public:
+  LatchEngage();
+
+  void Initialize() override;
+  
+  private:
+  Shooter* mShooter;
```
Comment (duane701) [inline src/main/include/commands/LatchEngage.h]: Subsystem by reference

```suggestion
  Shooter& mShooter;
```

File: src/main/cpp/subsystems/Shooter.cpp [hunk 15]
```diff
@@ -25,8 +25,24 @@ Shooter::Shooter(
...
+    shooterright.Follow(shooterleft);
+    SetPID();
 }
 
+void Shooter::SetPID()
+{
+  mshooterleft.Config_kP(0, p, 10);
+  mshooterleft.Config_kI(0, i, 10);
+  mshooterleft.Config_kD(0, d, 10);
+  mshooterleft.Config_kF(0, f, 10);
```
Comment (duane701) [inline src/main/cpp/subsystems/Shooter.cpp]: Prefix constants with a `k`. E.g `kP`, `kI`, etc.

File: src/main/include/subsystems/Shooter.h [hunk 16]
```diff
@@ -44,4 +46,6 @@ class Shooter : public frc2::SubsystemBase {
 
   frc::DoubleSolenoid& mhood;
   frc::DoubleSolenoid& mlatch;
+
+  double p, i, d, f;
```
Comment (duane701) [inline src/main/include/subsystems/Shooter.h]: pidf are all constants. Put these in the .cpp at the top in the `namespace {` block.

File: src/main/cpp/subsystems/Shooter.cpp [hunk 17]
```diff
@@ -23,6 +23,7 @@ Shooter::Shooter(
 , mshooterright{shooterright}
 , mhood{hood}
 , mlatch{latch}
+, p(0), i(0), d(0), f(0)
```
Comment (duane701) [inline src/main/cpp/subsystems/Shooter.cpp]: This goes away when these are `constexpr`

**Review by duane701 — CHANGES_REQUESTED** [review]

**Comment by duane701** [comment]
LGTM -- Rebase

**Comment by KenyonSchep** [comment]
Replaced by #30

**Pull request closed without merging**